--- 04-acceptors.js	2019-07-17 11:01:59.000000000 -0400
+++ 05-acceptors.js	2019-07-17 11:02:38.000000000 -0400
@@ -1,5 +1,45 @@
 /* global MeiosisRouting */
-const { routeTransition } = MeiosisRouting.state;
+const {
+  findRouteSegment,
+  routeTransition
+} = MeiosisRouting.state;
+
+import { Route, navTo } from "./05-routes";
+
+export const loginAccept = state => {
+  const currentLogin = findRouteSegment(
+    state.route.current,
+    "Login"
+  );
+  const previousLogin = findRouteSegment(
+    state.route.previous,
+    "Login"
+  );
+
+  if (
+    !currentLogin &&
+    previousLogin &&
+    !state.user &&
+    (state.login.username || state.login.password) &&
+    !confirm("You have unsaved data. Continue?")
+  ) {
+    return navTo([previousLogin]);
+  }
+};
+
+export const settingsAccept = state => {
+  if (
+    findRouteSegment(state.route.current, "Settings") &&
+    !state.user
+  ) {
+    return navTo([
+      Route.Login({
+        message: "Please login.",
+        returnTo: Route.Settings()
+      })
+    ]);
+  }
+};
 
 export const routeAccept = state => ({
   route: routeTransition(state.route)
